Sources/Regex/Capture.swift

```diff
@@ -42,11 +42,10 @@ extension AST {
     switch self {
     case let .alternation(child), let .concatenation(child):
       return child.any(\.hasCaptures)
-    case .capturingGroup, .group:
-      return true
-    case let .many(child), let .zeroOrOne(child), let .oneOrMore(child),
-         let .lazyMany(child), let .lazyOneOrMore(child),
-         let .lazyZeroOrOne(child):
+    case let .group(g, child), let .groupTransform(g, child, _):
+      return g.isCapturing || child.hasCaptures
+      || true // wait, why?
```

Comment:
@rxwei I'm doing this to preserve behavior. Two questions:

1) Why does compilation wrap concatenations in non-capturing groups if any child captures?
2) Why does it do it for non-capturing children too?

Reply:
> Why does compilation wrap concatenations in non-capturing groups if any child captures?

It actually wraps concatenations in a _capturing_ group if any child captures, just so that it gets pushed onto the stack and starts accumulating each capturing child's captures.  When we know that it has no captures, we skip that step.

> Why does it do it for non-capturing children too?

Does the above answer answer this question?

Reply:
I'm more confused. Prior to this refactoring, my understand is that `group` is non-capturing in the AST.

From the parser:
```swift
    switch lexer.peek()?.kind {
    case .leftParen?:
      lexer.eat()
      var isCapturing = true
      if lexer.tryEat(.question) {
        try lexer.eat(expecting: .colon)
        isCapturing = false
      }
      let child = try parse()
      try lexer.eat(expecting: .rightParen)
      return isCapturing ? .capturingGroup(child) : .group(child)
```

From the compiler:

```swift
    case .group(let child):
      instructions.append(.beginGroup)
      compileNode(child)
      instructions.append(.endGroup)
      return

    case .capturingGroup(let child, let transform):
      instructions.append(.beginCapture)
      compileNode(child)
      instructions.append(.endCapture(transform: transform))
      return

    case .concatenation(let children):
      let childrenHaveCaptures = children.any(\.hasCaptures)
      if childrenHaveCaptures {
        instructions.append(.beginGroup)
      }
      children.forEach { compileNode($0) }
      if childrenHaveCaptures {
        instructions.append(.endGroup)
      }
      return
```

But then in RECode we see something different:

```swift
    /// Begin a capture group.
    case beginGroup

    /// Ends a capture group.
    case endGroup

    /// Begin capturing a portion of the input string.
    case beginCapture

    /// End capturing a portion of the input string, transforming the substring with the specified
    /// transform.
    case endCapture(transform: CaptureTransform? = nil)
```

Reply:
Ah, that is a naming inconsistency.  `beginGroup` should have been named something like `beginTuple`. What it does is that it intercepts its children’s captures and joins them into a tuple. `beginCapture` on the other hand is more tied to what a literal capturing group does where it stores a substring matching the child.